ee/hogai/graph/insights/nodes.py

```diff
@@ -242,62 +307,254 @@ def _format_insights_page(self, page_number: int) -> str:
+            if series:
+                events = []
+                for s in series:
+                    if isinstance(s, dict):
+                        event_name = s.get("event", s.get("name", "Unknown"))
+                        events.append(event_name)
+                if events:
+                    # Limit to first 3 for LLM context window
+                    info_parts.append(f"Events: {', '.join(events[:3])}")
+
+            # Get date range info
+            date_range = query_dict.get("dateRange", {})
+            if date_range:
+                date_from = date_range.get("date_from", "")
+                if date_from:
+                    info_parts.append(f"Period: {date_from}")
+
+            return " | ".join(info_parts) if info_parts else None
+
+        except Exception:
+            return "Query error"
+
+    def _process_insight_for_evaluation(self, insight: Insight, query_executor: AssistantQueryExecutor) -> dict:
+        """
+        Process an insight for evaluation: convert to query, execute it, and create visualization message.
+        """
+        insight_info = {
+            "name": insight.name or insight.derived_name or "Unnamed",
+            "insight_id": insight.id,
+            "description": insight.description or "",
+            "query": "",
+            "filters": insight.filters or "",
+            "results": "",
+            "visualization_message": None,
+        }
+
+        try:
+            query_dict = None
+            query_kind = None
+
+            # If we have a query, use it directly
+            if insight.query:
+                if isinstance(insight.query, str):
+                    query_dict = json.loads(insight.query)
+                elif isinstance(insight.query, dict):
+                    query_dict = insight.query
+
+                if query_dict:
+                    query_kind = query_dict.get("kind")
+
+            if query_dict and query_kind:
+                if query_kind == "DataVisualizationNode":
+                    source = query_dict.get("source")
+                    if source and source.get("kind") == "HogQLQuery":
+                        query_dict = source
+                        query_kind = "HogQLQuery"
+
+                if query_kind in MAX_SUPPORTED_QUERY_KIND_TO_MODEL:
+                    # Execute query
+                    try:
+                        QueryModel = MAX_SUPPORTED_QUERY_KIND_TO_MODEL[query_kind]
+                        query_obj = QueryModel.model_validate(query_dict)
+                        results, _ = query_executor.run_and_format_query(query_obj)
+
+                        insight_info["query"] = (
+                            json.dumps(query_dict) if isinstance(query_dict, dict) else str(query_dict)
+                        )
+                        insight_info["results"] = results
+                    except Exception as e:
+                        insight_info["query"] = f"Failed to execute query: {str(e)}"
```

Comment:
Nitpick: Could it leak technical implementation through `e`? Not sure we care though, as open source.